File: tropirag/src/tropirag/governance/investigation.py
"""Descripteur de gouvernance — investigation MEDISUITE-CI-01, verrou M+18.

Source de vérité unique de l'état réglementaire du module CDS :

    « Sorties IA NON VALIDÉES cliniquement — investigation R6-R8 en cours,
      verrou M+18 ; toute décision clinique sur ces sorties est interdite. »

La phrase est identique au bandeau portail (clé i18n scr.ai_banner) — un
seul libellé réglementaire, affiché et appliqué. Le descripteur est porté
par des constantes immuables ; le mode `certified` (après marquage CE)
exige un opt-in explicite à double clé d'environnement, sinon le verrou
reste posé (fail-closed : toute valeur inconnue → locked).

Calendrier M+ : M0 est le jalon de déploiement de l'investigation ; les
fenêtres R5→R8 (plan de validation v1.0.0) sont exprimées en mois pleins
depuis M0. Le verrou de base M+18 clôt R6 (inclusions + monitoring) ; il
est suivi de R7 (rapport clinique MEDDEV 2.7/1) puis R8 (notifié, CE).
"""
from __future__ import annotations

import os
from datetime import date
from typing import Any

__all__ = [
    "GOVERNANCE_DESCRIPTOR",
    "PROTOCOL",
    "PHASES_IN_PROGRESS",
    "LOCK_LABEL",
    "BANNER_TEXT",
    "FORBIDDEN_ACTION",
    "DECISION_PATHS",
    "governance_mode",
    "is_decisional_use_allowed",
    "stamp",
    "add_months",
    "m18_date",
    "m18_status",
    "finalize_denial",
]

# --- descripteur réglementaire (immuable) ----------------------------------

PROTOCOL = "MEDISUITE-CI-01"

PHASES_IN_PROGRESS: tuple[str, ...] = ("R6", "R7", "R8")

LOCK_LABEL = "M+18"

#: Libellé unique — identique à la clé i18n portail `scr.ai_banner` (4 langues).
BANNER_TEXT = (
    "Sorties IA NON VALIDÉES cliniquement — investigation R6-R8 en cours, "
    "verrou M+18 ; toute décision clinique sur ces sorties est interdite."
)

FORBIDDEN_ACTION = "toute décision clinique sur ces sorties est interdite"

_REFS = (
    "MDR (UE) 2017/745 Annexe XV — investigation clinique",
    "ISO 14155 — bonnes pratiques cliniques",
    "MEDDEV 2.7/1 rev 4 — évaluation clinique (R7, ADR-0026)",
    "Protocole TD-10 v1.0 — compliance/mdr/technical-documentation/"
    "10-protocole-investigation-multicentrique-R5.md",
)

#: Chemins API dont la réponse matérialise une sortie d'investigation
#: (tampon `governance` injecté dans le corps + en-têtes).
DECISION_PATHS: tuple[str, ...] = (
    "/api/v1/cases",
    "/api/v1/clinical",
    "/api/v1/evidence",
    "/api/v1/surveillance",
)


def governance_mode() -> str:
    """Mode effectif — fail-closed : `certified` exige le double opt-in
    (MEDISUITE_GOVERNANCE_MODE=certified ET MEDISUITE_GOVERNANCE_CE_ACK=ce),
    toute autre valeur (y compris inconnue) retombe sur `locked`."""
    mode = (os.getenv("MEDISUITE_GOVERNANCE_MODE") or "").strip().lower()
    if mode == "certified" and (os.getenv("MEDISUITE_GOVERNANCE_CE_ACK") or "").strip() == "ce":
        return "certified"
    return "locked"


def is_decisional_use_allowed() -> bool:
    """True uniquement après marquage CE effectif (double opt-in)."""
    return governance_mode() == "certified"


def stamp() -> dict[str, Any]:
    """Bloc `governance` injecté dans les sorties décisionnelles."""
    allowed = is_decisional_use_allowed()
    return {
        "statut": "certified" if allowed else "investigation",
        "protocole": PROTOCOL,
        "phases_en_cours": [] if allowed else list(PHASES_IN_PROGRESS),
        "verrou": LOCK_LABEL,
        "decision_clinique": "autorisée" if allowed else "interdite",
        "avis": BANNER_TEXT,
        "references": _REFS,
    }


GOVERNANCE_DESCRIPTOR = stamp()


# --- calendrier M+ (mois pleins depuis M0) ----------------------------------


def add_months(d: date, months: int) -> date:
    """Décalage calendaire en mois pleins (clamp fin de mois : 31/02 → 28/02)."""
    y = d.year + (d.month - 1 + months) // 12
    m = (d.month - 1 + months) % 12 + 1
    day = min(d.day, [31, 29 if y % 4 == 0 and (y % 100 != 0 or y % 400 == 0) else 28,
                      31, 30, 31, 30, 31, 31, 30, 31, 30, 31][m - 1])
    return date(y, m, day)


def m18_date(m0: date) -> date:
    """Date du verrou de base M+18 (clôture R6 — inclusions + monitoring)."""
    return add_months(m0, 18)


def m18_status(today: date, m0: date) -> dict[str, Any]:
    """État du verrou M+18 pour `today` — phase dérivée SANS horloge mondiale
    (réplique la logique portail features/study/status-logic.ts activePhase) :
    verrou non posé → R6 (inclusions/monitoring) ; posé → R7 (rapport clinique) ;
    l'entrée en R8 est une décision humaine (notifié) — non dérivable.
    Le mode `certified` (double opt-in) est l'unique sortie du régime."""
    m18 = m18_date(m0)
    locked = today >= m18
    allowed = is_decisional_use_allowed()
    phase = "R7" if locked else "R6"
    months_elapsed = (today.year - m0.year) * 12 + (today.month - m0.month)
    if today.day < m0.day and months_elapsed > 0:
        months_elapsed -= 1
    return {
        "m0": m0.isoformat(),
        "verrou_m18": m18.isoformat(),
        "pose": locked,
        "phase_active": phase,
        "phases_en_cours": list(PHASES_IN_PROGRESS) if not allowed else [],
        "mois_ecoules": months_elapsed,
        "jours_avant_verrou": max(0, (m18 - today).days),
        "decision_clinique": "autorisée" if allowed else "interdite",
    }


# --- garde fail-closed (451) ------------------------------------------------


def finalize_denial(case_id: str | None = None) -> dict[str, Any]:
    """Corps structuré du refus 451 — matérialisation de l'interdiction.
    `451 Unavailable For Legal Reasons` : la non-disponibilité est LEGALE
    (régime d'investigation), pas technique."""
    body = {
        "error": "clinical_decision_locked",
        "status_code": 451,
        "governance": stamp(),
        "explication": (
            "La matérialisation d'une décision clinique à partir d'une sortie "
            "TropiRAG est interdite tant que l'investigation MEDISUITE-CI-01 "
            "n'est pas close (rapport clinique R7 + marquage CE R8). Les "
            "sorties restent consultables à titre d'investigation uniquement."
        ),
        "reprise": "après marquage CE (fenêtre R8, M+21 → M+30+) + double opt-in",
        "journal": "audit:governance.decision.denied",
    }
    if case_id:
        body["case_id"] = case_id
    return body


# --- m0 par défaut (simulation) ---------------------------------------------

#: Jalon de déploiement de l'investigation (M0) — surcharge via
#: MEDISUITE_GOVERNANCE_M0 (ISO 8601). Défaut aligné sur la simulation
#: du plan de validation : 2025-06-01 → verrou M+18 le 2026-12-01.
def default_m0() -> date:
    raw = (os.getenv("MEDISUITE_GOVERNANCE_M0") or "").strip()
    if raw:
        try:
            return date.fromisoformat(raw)
        except ValueError:
            pass
    return date(2025, 6, 1)

File: apps/web-portal/src/features/governance/investigation.ts
/**
 * Gouvernance clinique — investigation MEDISUITE-CI-01, verrou M+18 (v0.14).
 *
 * Miroir frontend du module backend `tropirag.governance.investigation` :
 * UN libellé réglementaire (identique à la clé i18n `scr.ai_banner`, déjà
 * affichée sur les 28 écrans AiAssist) et UN calendrier M+ pour le compte à
 * rebours du verrou de base. Logique pure, sans réseau ni store — le
 * composant GovernanceBanner ne fait que l'affichage.
 *
 * Phases (plan de validation v1.0.0) : R6 inclusions+monitoring → VERROU DE
 * BASE M+18 → R7 rapport clinique (MEDDEV 2.7/1) → R8 notifié + marquage CE.
 * Tant que l'investigation est ouverte, toute décision clinique fondée sur
 * une sortie IA est interdite — l'interdiction est aussi appliquée côté API
 * (tampon `governance` + garde 451 sur POST /api/v1/decision/finalize).
 */

export const PROTOCOL = "MEDISUITE-CI-01";

/** Phases d'investigation restantes avant usage décisionnel. */
export const PHASES_IN_PROGRESS = ["R6", "R7", "R8"] as const;

export const LOCK_LABEL = "M+18";

/** Phrase réglementaire — doit rester identique à `scr.ai_banner` (i18n). */
export const BANNER_TEXT =
  "Sorties IA NON VALIDÉES cliniquement — investigation R6-R8 en cours, verrou M+18 ; toute décision clinique sur ces sorties est interdite.";

/** Tampon apposé sur CHAQUE sortie IA restituée (résultats, synthèses). */
export const OUTPUT_STAMP = "Sortie d'investigation — non décisionnelle";

/** Jalon de déploiement de l'investigation (M0) — aligné sur le backend
 * (tropirag.governance.investigation.default_m0, surcharge M0 côté serveur). */
export const M0: Readonly<Date> = new Date(2025, 5, 1); // 2025-06-01

/** Décalage calendaire en mois pleins (clamp fin de mois : 31/02 → 28/02). */
export function addMonths(d: Date, months: number): Date {
  const y = d.getFullYear() + Math.floor((d.getMonth() + months) / 12);
  const m = (d.getMonth() + months) % 12;
  const daysInMonth = new Date(y, m + 1, 0).getDate();
  return new Date(y, m, Math.min(d.getDate(), daysInMonth));
}

/** Date du verrou de base M+18 (clôture R6 — inclusions + monitoring). */
export function m18Date(m0: Readonly<Date> = M0): Date {
  return addMonths(new Date(m0), 18);
}

const DAY_MS = 86_400_000;

function startOfDay(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), d.getDate());
}

/** Différence en jours pleins (a − b). */
export function daysBetween(a: Readonly<Date>, b: Readonly<Date>): number {
  return Math.round((startOfDay(new Date(a)).getTime() - startOfDay(new Date(b)).getTime()) / DAY_MS);
}

export interface M18Status {
  /** Verrou de base posé (date atteinte). */
  pose: boolean;
  /** Phase dérivée — verrou non posé → R6 ; posé → R7 ; R8 = décision
   * humaine (notifié), non dérivable (cf. features/study/status-logic.ts). */
  phaseActive: "R6" | "R7";
  /** Date ISO du verrou M+18. */
  verrouIso: string;
  /** Jours restants avant M+18 (0 si posé). */
  joursAvantVerrou: number;
  /** Mois pleins écoulés depuis M0. */
  moisEcoules: number;
}

/** État du verrou M+18 pour `today` — réplique `m18_status` backend. */
export function m18Status(today: Readonly<Date> = new Date(), m0: Readonly<Date> = M0): M18Status {
  const m18 = m18Date(m0);
  const pose = startOfDay(new Date(today)).getTime() >= startOfDay(new Date(m18)).getTime();
  const moisEcoules =
    (today.getFullYear() - m0.getFullYear()) * 12 + (today.getMonth() - m0.getMonth()) -
    (today.getDate() < m0.getDate() ? 1 : 0);
  return {
    pose,
    phaseActive: pose ? "R7" : "R6",
    verrouIso: m18.toISOString().slice(0, 10),
    joursAvantVerrou: Math.max(0, daysBetween(m18, today)),
    moisEcoules: Math.max(0, moisEcoules),
  };
}

/** Ligne de compte à rebours du bandeau (fr, cohérente avec le portail). */
export function countdownLabel(st: M18Status = m18Status()): string {
  const dateFr = new Date(st.verrouIso + "T00:00:00").toLocaleDateString("fr-FR", {
    day: "numeric", month: "long", year: "numeric",
  });
  return st.pose
    ? `Verrou de base M+18 posé (${dateFr}) — phase active ${st.phaseActive} ; usage décisionnel toujours interdit jusqu'au marquage CE.`
    : `Phase active ${st.phaseActive} · verrou de base M+18 : ${dateFr} (dans ${st.joursAvantVerrou} j)`;
}

/** Bloc `governance` renvoyé par l'API TropiRAG (tampon des sorties CDS). */
export interface GovernanceStamp {
  statut: "investigation" | "certified";
  protocole: string;
  phases_en_cours: string[];
  verrou: string;
  decision_clinique: "interdite" | "autorisée";
  avis: string;
}

/** Le tampon reçu d'une sortie CDS est-il conforme au régime attendu ? */
export function stampIsInvestigational(st: GovernanceStamp | undefined | null): boolean {
  return !!st && st.statut === "investigation" && st.decision_clinique === "interdite";
}
